RFC - Localizing API (Backend) messages (#3)
folio-org/rfcs: text/0000-localizing-api-messages.md

> - Start Date: 03-29-2022

Reviewer: When using hyphens the usual format is yyyy-mm-dd.

- RFC PR: 
- FOLIO Issue: 
- Submitter : Radhakrishnan Gopalakrishnan (rgopalakrishnan@ebsco.com)
- Co-Submitter(s) : Zak Burke (zburke@ebsco.com)
- Sub Group :

# Localizing API (Backend) Messages

## Summary
  The purpose of this RFC is to define an approach that allows API developer to provide
messages in the API response depending on the language preference set in the request. 

## Motivation

- Help users understand the information coming from the application backend using a language they understand better
- To allow FOLIO platform to be used by users across the world 

> - To create a tremendous opportunity for growth that we can never could have achieved within just one country.

Reviewer: Help translators by standardizing keys and syntax (plural syntax, etc.).
Help translation devops by standardizing the location of translation files in the repositories and their format.

Author: There was a debate initially about the scope of this RFC and the level on detail/directions we needed to provide. So, interest of minimizing the scope and being less prescriptive about implementation, these were left out. I certainly amenable to adding this back in.

Author: Added

### Terminology

> - Static Message - E.g. _This is something you need translate_

Reviewer 2: > When going through JIRA, I found references to messages with line breaks, messages in HTML format etc.. I called these out mainly to highlight the type of messages this RFC handles. Messages in its most basic forms.

I appreciate the thorough analysis of the wide array of JIRA issues. I am a little concerned that this could lead to us combining all of those sources of information together in a way that doesn't really express the intended whole. 

Maybe these could be combined together - `static message with placeholder(s)` automatically covers static messages without placeholders.

And that statement could be moved to the section that describes the scope of this RFC.

> - Static Message with placeholder - _You can ONLY renew {n} times_

Reviewer: Is the name field of a reference data record a static message?

Author: From what I understand, values in the database are referred to as controlled vocabulary. So, the name column value from a DB record falls into that category

Reviewer 2: > From what I understand, values in the database are referred to as controlled vocabulary. 

Please be careful using the term Controlled Vocabulary. It has a specific meaning within the library domain (to do with how change management for those values work) 

Not all reference records are controlled vocabulary.

> So, the name column value from a DB record falls into that category

Which category are you referring to?

As I understand it, the names of reference records are considered `dynamic` values. @zburke I think you coined that terminology originally, how wrong am I?

Author: If controlled vocabulary has no relevance when it comes to localization, I will replace it with "Dynamic messages", to refer to the data in the DB that needs to be localized.

Reviewer 2: > If controlled vocabulary has no relevance when it comes to localization, I will replace it with "Dynamic messages", to refer to the data in the DB that needs to be localized.

I think replacing it with dynamic values is a good idea.

 

### In Scope Requirements/Use cases
- Return localized messages based on the value passed in the accept-language header
- Handle static messages, static messages with placeholders

> ### Out of Scope Requirements/Use cases

Reviewer: We already have a decision whether messages should be generated by the back-end module itself or at some other module: https://wiki.folio.org/display/DD/Handle+i18n+where+messages+are+generated
Therefore this decision should be listed as out-of-scope of this RFC.

Reviewer 3: You're right, @julianladisch, this RFC covers a lot of the same territory as the RFC. Since tech leads does not meet any longer, the decision log hasn't been well maintained. For architectural decisions like how to manage i18n, we hope the RFC process can be a way to make such decisions in the future and to document them. 

Part of the goal here is to coalesce the i18n info that is scattered across Jira and the decision log and TC meeting notes, and part of it is to pilot the RFC process with a topic that is (hopefully) relatively uncontroversial and already settled, if not completely fleshed out. If we pushed these decisions -- use the `accept-languages` header, handle static values -- out of scope, both goals would suffer. Thus, I think this should stay, even if it covers some existing territory.

> ### Out of Scope Requirements/Use cases

Reviewer: We already have a decision how to pass the requested language: https://wiki.folio.org/display/DD/Localization+parameter+for+back-end
Therefore this decision should be listed as out-of-scope of this RFC.

> - [Plural Syntax](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Pluralsyntax)

Reviewer 4: Why is plural syntax out-of-scope?  Is the assumption that a plural syntax won't be needed?  Is there a reason not to use the [ICU format](https://lokalise.com/blog/complete-guide-to-icu-message-format/) that the front-end modules already use?

Author: I understand that it is relevant.  It was mainly left out because I was trying to limit the scope. If we need to include it, we also need to start considering similar requirements that has been documented in JIRA. Perfectly Ok to include it if there is an agreement

Reviewer 4: Ah, I see.  Was the RFC written from a perspective that the FOLIO project would need to write its own localization library?  I ask because ICU libraries seem to be pretty well supported ([ICU4J](https://unicode-org.github.io/icu/userguide/icu4j/)), and the ICU format is what the front-end developers are already using.

Author: I expect the developers to write their own library ONLY when they have exhausted all available options provided by the language and the frameworks we use.  Regarding ICU4J, I am not saying that we shouldn't use it.  I would like us to compare it against the support provided by java and spring. If there are gaps in what it offers, we can certainly use that as a justification for introducing ICU4j. I left these out from the RFC mainly to avoid being very prescriptive about tool/framework choice.

Reviewer 2: > It was mainly left out because I was trying to limit the scope. If we need to include it, we also need to start considering similar requirements that has been documented in JIRA.

I think we need to be careful about what scope we choose to exclude.

Changing the format of the message templates or the files that include them after this RFC has been approved and adopted could be a significant endeavour affecting multiple teams.

Reviewer 3: Agreed, @dltj. I see two options here: 
1. we say message format is out of scope and punt it to a separate RFC so we don't get bogged down here. 
2. we say message format is in scope and the format shall be ICU if there is unanimous agreement. 

Personally I think we'd be well-served by (2); ICU is well established and documented and it behooves us, as it does any project, to follow this industry-standard practices. If folks think this is worth discussion, however, I think we should move it to a separate RFC so as not to delay this one.

Reviewer: If message format is in scope then we need a spike/proof-of-concept for each option. This may take longer than the [two weeks of the draft review](https://wiki.folio.org/pages/viewpage.action?spaceKey=~vijay&title=RFC+Process).

> - Naming convention to be used for keys

Reviewer: Translators need good keys so that they can find where and how the message is used. Some messages are ambiguous and require a good key when translating.
Therefore this needs to be addressed. Either in this RFC or in parallel RFC.

Author: Ok to include key format in scope if there is consensus

Author: I think we agreed to have the keys as strings and not provide further guidance as it may conflict with conventions that are already in place.

> - Process for managing (decoupled from lokalise.com) translations

Reviewer 4: Respectfully, if Lokalise is needed to manage the translations, then the process for managing the translations with Lokalise is in scope.  If Lokalise is not needed, then this RFC needs to define a process by which translators will add the translated strings.

It might be possible to use the existing Lokalise→GitHub script with a few different parameters in the API call to Lokalise.  But this would need to be tested.

Reviewer 4: I'm withdrawing this comment after watching the [TC meeting recording](https://recordings.openlibraryfoundation.org/folio/tech-council/2022-04-20T11:00/) and listening to @zburke's comments.  The use of a standard message format (like the ICU message format comment above) is even _more_ important, though, if there is a desire to decouple the architecture from the implementation mechanism.

- Returning formatted (HTML/Markdown) messages

> - Usage of [soft hyphen](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Softhyphentobreakwords)to break messages

Reviewer: ```suggestion
- Usage of [soft hyphen](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Softhyphentobreakwords) to break messages
```

- Support customization per tenant

> - Process for back porting API/Backend messages similar to what we have for the front end 

Reviewer: Thanks, can you put the link into the document, similar to the soft hyphen link?

> - Support for controlled vocabulary (runtime data/data coming from DB tables. E.g. Patron Groups for tenants)

Reviewer 4: Add reference to [UXPROD-3148](https://issues.folio.org/browse/UXPROD-3148), please.

## Detailed Explanation/Design

> - Locale specific translations MUST be stored in a properties file

Reviewer 2: Are property files specific to Java?

Author: Every framework/language has a way of loading config info using a generic key value pair format. So, I wouldn't say it is java specific. That said, along with loading config info using this format, in java, language specific resource bundles are specified in this format to load language translations.

Reviewer 2: > Every framework/language has a way of loading config info using a generic key value pair format. So, I wouldn't say it is java specific. 

I would imagine that each of those languages will do that slightly differently, even if it's only in naming, location and minor format changes e.g. a version header.

> That said, along with loading config info using this format, in java, language specific resource bundles are specified in this format to load language translations.

I think it is important that we separate the general protocol decision from the language specific implementation guidance.

FOLIO has a history of mixing the two together which has led to folks misunderstanding what it is they are expected to do.

Author: The format is as simple as 

a=b
c=d

The file name could be anything.

So, instead of using the properties file, I could refer them as key value pairs and use .properties as an example for Java. Will that be better?

Reviewer 2: > So, instead of using the properties file, I could refer them as key value pairs and use .properties as an example for Java. Will that be better?

Yes, I think splitting the Java specific standard (because it says MUST) from the parts that are intended to be general is a good idea.

Personally (and based upon the TC conversation, other folks don't support this) I consider these files to be part of an interface between the module source and the translations process.

A part of the reason why I'm concerned about this aspect is that I think the decision is based upon it being convenient for Java modules. Much like I imagine the choice of JSON for the UI (JS based) modules was convenient for them.

Assuming this goes through, we will have two different file formats for translations. Meaning whatever part of the system / operations / infrastructure / community need to interact with them have to accommodate both of those formats.

It wouldn't surprise me if the next time we accept a new language / runtime as common into FOLIO, we'll end up with a third format because it was convenient for that environment too.

We already have back end modules that use other tools (node.js, Perl, grails) and I imagine we are expecting those modules to comply with the standards outlined here as well.

> ## Detailed Explanation/Design

Reviewer: I don't comment on this section because at this stage of the RFC process we are asked to only comment on the scope, not on the content.

Reviewer 2: In a [thread](https://folio-project.slack.com/archives/CAQ7L02PP/p1650473930895659?thread_ts=1650471081.413709&cid=CAQ7L02PP) on Slack, @vijay-eis suggested to me that all feedback is acceptable at this point in the process.

Author: @marcjohnson-kint You are correct. The RFC in its entirety is under review

- Properties file naming convention
  - default locale (en_US) - somename.properties
  - specific locale - somename_fr.properties (For french),  somename_de.properties (For german)

> - Properties file MUST BE loaded using the classpath. AVOID using file locations

Author: I wanted to standardize on the mechanism that was used load these files without adding additional complexity. Had I suggested that we use file locations, I need to specify where these will be located. Absolute vs relative path. If it is absolute, how will I pass on this information to the code that needs it. System properties, env var etc.. If it is relative, relative to what. Wanted to avoid all these complexities :)

> - Properties file MUST BE loaded using the classpath. AVOID using file locations

Reviewer 2: > I wanted to standardize on the mechanism that was used load these files without adding additional complexity. 

I understand not wanting to add unnecessary complexity.

> Had I suggested that we use file locations, I need to specify where these will be located. Absolute vs relative path. If it is absolute, how will I pass on this information to the code that needs it. System properties, env var etc.. If it is relative, relative to what. Wanted to avoid all these complexities :)

I agree that for Java loading via the class path is better than file paths.

I think that we need to acknowledge that this has trade offs in complexity in other ways:
* Folks need to understand the guidance is Java (or at least JVM specific)
* Folks need to understand that the location of these files might vary between modules
* The operational processes for managing translations need to accommodate these variations

> - Properties file MUST BE loaded using the classpath. AVOID using file locations

Reviewer: How properties get loaded is an implementation detail that should be out of scope of an RFC. If there are best practices they can be published on dev.folio.org or wiki.folio.org.

The translation workflow might be out of scope. However, each translation workflow need to pick up some files from the GitHub repository and need to store translated files in the GitHub repository. Defining a standard location will help the translation workflow devops and the translators. Therefore this interface should be in scope and been standardized.

> - Properties file MUST BE loaded using the classpath. AVOID using file locations

Author: I think we agreed to recommend that language/framework specific best practices MUST be followed when loading translation files at runtime

- Properties file entries WILL use the standard key value pair format

> - When using application frameworks (spring, vertx, etc.) implement localization using the framework specific implementation (if there is one)

Reviewer 2: I think it is important to acknowledge the trade off that we are introducing with this decision.

When this was discussed during the TC meeting, an impact of this was that the translation files could be in different places within the source code repository.

I understand that it makes it easier for developers, because folks are not working against the chosen tooling.

It creates a non-uniform interface to the outside world that needs to change the translations. That means we are putting the burden on the operations tooling and processes to accommodate this.

My bias tends to be the other way, we accept greater internal complexity (e.g. configuration of frameworks) to present less external complexity (a uniform location for translation property files).

I think this exposes our build (and localization) tooling to the potential of continual changes e.g. any time folks use a different language / framework (which some folks advocate for, we already have Vert.x, RMB, Spring Way, Grails, Node.js and Perl).

It undermines the idea that a module should comply with a set of expectations irrespective of the tooling choices.

(This is based upon the assumption that localization is one of those expectations, which I think is the intent of this proposal)

Reviewer 2: @zburke 

> you're right, there's a "burden" to be carried here. I think placing that burden "on the operations tooling and processes" is exactly the right place for it. It consolidates this complexity (probably to a single script that syncs the values in our repos with the values in Lokalise or Smartling or Translized or whatever) instead of spreading it across every single repository. That seems like a great trade off to me.

I think making that decision and the trade offs of that approach explicit and prominent in this RFC would be great, to help both folks reviewing this now and folks using this as documentation in the future to understand this aspect.

I think that is important because this decision is broad in scope and impact, given that it affects how uniform the interface modules expose to the rest of the project are and has implications for build tooling and integration with services like Localize.

That could be even more important as FOLIO potentially moves towards a model where modules might be supported by a wider audience and are hosted elsewhere yet also need to integrate with these shared services (translation change management).

If I were to reflect my understanding I might express it something like the following.

Decision: each language / runtime MAY to use different file formats for static translations
Trade offs:
* allows each language / runtime variation to use more easily use existing tooling for localization
* Reduces the complexity of modules and effort for development teams as no mapping between external expectations and internal tool choices needs to be done
* Increases complexity of the builds which are involved in the centralised translation change management process (currently using Lokalise) and effort for DevOps folks

These trade offs assume that the tooling choice and structures will be standardised for each language / runtime rather than on a per module basis.

For that to be the case, we would need at a minimum, to define those standards for:
* Perl / Make
* Grails / Gradle
* Node.js / Yarn

- When handling messages with placeholders, replace the placeholders with the actual value on the server side before
  sending it back to the client
- STOP using the lang parameter.
- accept-language header value MUST be in ll_CC format. , where ll is a two-letter language code, 
  and CC is a two-letter country code.
- When accept-language header is missing or has an incorrect value, return the message in en_US (or the default locale configured)

> ## Rationale and Alternatives

Reviewer: Thank you for collecting and writing up this "Rationale and Alternatives" section.
We already have made a decision whether messages should be generated by the back-end module itself or at some other module: https://wiki.folio.org/display/DD/Handle+i18n+where+messages+are+generated
Therefore this decision is out-of-scope of this RFC.
This section "Rationale and Alternatives" belongs to that decision. I've copied this section over there: https://wiki.folio.org/display/DD/Handle+i18n+where+messages+are+generated
Can this section removed from this RFC?

*Cons*
- Additional overhead in managing more ui modules
- Not all originating messages can be mapped to a frontend module

### Option 3
The back-end module hosts the language files.

mod-circulation creates a translations/mod-circulation/ directory.

Stripes fetches the translation files from the back-end module and can process them the same way as translation files from frone-end modules.

*Pros*
- Software and language files are in the same repository. Options 1 and 2 require to add the message string to some other repository using a second pull request.

*Cons*
- Is this technically feasible? How can Stripes fetch languages files from a non-Stripes module?

### Option 4:
When calling the back-end pass the required locale. The back-end maintains the translations files, replaces the placeholders and puts the final string into the "message" property. A lang query parameter that many FOLIO API interfaces already have or an "accept-language" header line might be used.

*Pros*
- Java code and translation files live in the same repository. No work for the front-end.
- This is a well known pattern in the Java community

*Cons*

> - None

Reviewer 2: I think not having any trade offs for an option could be a hint that we are missing something.

I think one of the trade offs with this approach is that it means that FOLIO has two different ways of approach localization.

Author: I didn't list anything under cons largely due the information I have gathered based on my experience. That said, the information I have could very well be incomplete when it comes to FOLIO. We can certainly include trade offs if we can think of any.

Reviewer 2: > We can certainly include trade offs if we can think of any.

My starters for this could be:
* This standard (which is mixed together with the protocol decision) is Java / JVM specific
* There will be two (and likely a third to come when we consider dynamic values) separate ways for translations to be described (properties files) and managed (which will need to accommodate the format and location differences)

These likely overlap with @todolson comment below.

## Risks and Drawbacks

> None that we can think of. The pattern we are adopting is a standard pattern in the java community

Reviewer 2: Similarly to my comment about cons, it is unlikely that there are no drawbacks to an approach at all.

Is this the section to talk about the impact of not considering the process for changing the translations?

To me, that area presents the most potential for trade offs and drawbacks that don't seem to be mentioned in this document yet.

Reviewer 5: One drawback is that the translation proposal is specific to one programming language. If we have a long-term vision of multiple languages, that might be an issue and new conventions would need to be implemented. Another drawback I infer is that translators have to hunt down every module.

One option that has not been considered is a central service to store localization strings. That would be available to modules written in whatever programming language. As a side-effect, it would permit reusing message text across modules.

Conceivably, back-end _or_ front-end modules could call it with a key and a language and get back the needed string. A module could cache translations as they are called, and the service could issue an alert when there are new translations and the cache needs to be refreshed.

Translators have one central storage to see everything that needs to be translated. If a key were constructed via a naming convention, that could intentionally include info about the use context to translators.

Reviewer 3: Our [module submission guidelines state that "Back-end modules are based on Maven/JDK 11 and provide a Dockerfile"](https://github.com/folio-org/tech-council/blame/e72b82ba80aa51a5e8c5e989d2bafb86d7488b98/MODULE_EVALUATION_TEMPLATE.MD#L31) thus I see no problem with JDK-specific guidance here.

Reviewer 2: > Our [module submission guidelines state that "Back-end modules are based on Maven/JDK 11 and provide a Dockerfile"](https://github.com/folio-org/tech-council/blame/e72b82ba80aa51a5e8c5e989d2bafb86d7488b98/MODULE_EVALUATION_TEMPLATE.MD#L31) thus I see no problem with JDK-specific guidance here.

I think this might be ones of those cases where we accepted some limitations in the criteria due to it only applying to  new modules and the time constraints we had to put this together.

Assuming this decision will apply to all existing modules, it will need to cater to (at least) the following languages / runtimes / build systems:
* java / maven
* node.js / yarn (mod-graphql)
* grails / gradle (mod-agreements, mod-licences etc)
* perl / make (net-Z39-50-FOLIO)

At the very least, I think it could be good to distinguish (e.g. separate sections / headings) the general standard from the language / runtime specific guidance.

## Unresolved Questions

> - How do we allow tenants to change default locale ?

Author: Default locale is the locale that will be used when the locale information is not available when loading the resource bundles. For US tenants, I expect this to be English and for a German tenant it should be German. Without setting a default value for the locale, it will default to what the JDK is using.  Person installing the JRE should set the appropriate locale.

> - How do we allow tenants to change default locale ?

Reviewer 2: > Default locale is the locale that will be used when the locale information is not available when loading the resource bundles.

My interpretation of this RFC is that the default locale (from the back end modules perspective) is always US English.

How is that default tenant specific?

> For US tenants, I expect this to be English and for a German tenant it should be German. Without setting a default value for the locale, it will default to what the JDK is using. Person installing the JRE should set the appropriate locale.

Oh, we are expecting that system operators change the configuration of the JRE to specify a locale.

This does not seem to be outlined in this RFC. Have I missed that?

As I understand it, that contradicts the principle that FOLIO module instances can be used by multiple tenants (each of which could operate a different locale).

> - How do we allow tenants to change default locale ?

Reviewer 3: > My interpretation of this RFC is that the default locale (from the back end modules perspective) is always US English.

Agreed. A module should not have any knowledge of a tenant's locale or its JRE's locale, only a request's locale. My understanding is that a module's HTTP response should reflect the HTTP request's `Accept-language` header, or `en-US` if no such header is present.

* Start Date: 2022-03-29
* RFC PR: 
* FOLIO Issue: 
* Submitter : Radhakrishnan Gopalakrishnan (rgopalakrishnan@ebsco.com)
* Co-Submitter(s) : Zak Burke (zburke@ebsco.com)
* Sub Group : 
  * Marc Johnson (marc.johnson@k-int.com)
  * Julian Ladisch (julian.ladisch@gbv.de)
  * Peter Murray (peter@indexdata.com)

>   * Tod Olson (tod@uchicago.edu)

Reviewer 2: I'm not sure we should classify folks who provided feedback (and some who attended a meeting to discuss that feedback) as a working group.

- Handle static messages with placeholder(s)

### Out of Scope Requirements/Use cases
- [Plural Syntax](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Pluralsyntax)
- Naming convention to be used for keys
- Process for managing (decoupled from lokalise.com) translations
- Returning formatted (HTML/Markdown) messages
- Usage of [soft hyphen](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Softhyphentobreakwords) to break messages
- Support customization per tenant
- Process  for [back porting](https://wiki.folio.org/display/I18N/Backport) API/Backend messages similar to what we have for the front end 
- Support for [controlled vocabulary](https://issues.folio.org/browse/UXPROD-3148) (runtime data/data coming from DB tables. E.g. Patron Groups for tenants)

## Detailed Explanation/Design
* Translation file format will depend on the language that is used to develop the module
  * Java, Groovy - properties
  * JavaScript, nodejs - json
  * Any other language - Please contact Technical Council for further guidance
* Translation files MUST be placed under translations/\<Backend Module Name\>

> * Translation file name should be ll_CC.ext. Here ‘ll’ is an ISO 639 two-letter language code, and ‘CC’ is an ISO 3166 two-letter country code.

Reviewer 2: ```suggestion
* Translation file name SHOULD be [language code]_[country code].[extension] were `language code` is an ISO 639 two-letter language code, `country code` is an ISO 3166 two-letter country code and `extension` is the appropriate extension for the file format. For example, in `ll_CC.json`, ‘ll’ is the language code, ‘CC’ country code and `json` is the extension.
```

> * Translation file name should be ll_CC.ext. Here ‘ll’ is an ISO 639 two-letter language code, and ‘CC’ is an ISO 3166 two-letter country code.

Reviewer 2: We state MUST for other expectations. The use of should here (which I've capitalized in my other suggestion) suggests that there are legitimate exceptions to this rule. Is that the case?

* Translation message keys MUST be a string (alphanumeric characters ONLY)

> * Developer should follow language/framework specific recommended best practices for loading translation at runtime

Reviewer 2: ```suggestion
* Module SHOULD use language/framework specific recommended best practices for loading translation at runtime
```

* When handling messages with placeholders, replace the placeholders with the actual value on the server side before
  sending it back to the client
* STOP using the lang query string parameter.
* accept-language header value MUST be in ll_CC format. , where ll is a two-letter language code, 
  and CC is a two-letter country code.
* When accept-language header is missing or has an incorrect value, return the message value in en_US. If message value 

>   cannot be found, return the message key

Reviewer 2: I think it could make it easier to follow the guidance if it was split into 3 sections (please feel free to come up with better names):
* API protocol
* Translation definitions
* Language / Runtime specific

* Start Date: 2022-03-29
* RFC PR: 
  * Preliminary Review: https://github.com/folio-org/rfcs/pull/3
  * Draft Review :
  * Public Review :
  * Final Review :
* FOLIO Issue:
* Current Status: :no_entry_sign: DRAFT REVIEW :no_entry_sign:
* Submitter : Radhakrishnan Gopalakrishnan (rgopalakrishnan@ebsco.com)
* Co-Submitter(s) : Zak Burke (zburke@ebsco.com)
* Sub Group : 
  * Marc Johnson (marc.johnson@k-int.com)
  * Julian Ladisch (julian.ladisch@gbv.de)
  * Peter Murray (peter@indexdata.com)
  * Tod Olson (tod@uchicago.edu)

# Localizing API (Backend) Messages

## Summary

>   The purpose of this RFC is to define an approach that allows API developer to provide
> messages in the API response depending on the language preference set in the request. 

Reviewer 2: ```suggestion
The purpose of this RFC is to define an approach that allows API developer to provide
messages in the API response depending on the language preference set in the request. 
```

## Motivation

- Help users understand the information coming from the application backend using a language they understand better
- To allow FOLIO platform to be used by users across the world 
- To create a tremendous opportunity for growth that we can never could have achieved within just one country.
- To Promote consistency and make it easier for tools and translators when handling translation files

### Terminology
* Base Language - Language on which all other translations are based.
* Message Key - Refers to the string that is used in place of a hardcoded message in the code
* Message Value - Refers to the string that is fetched using a message key lookup for a given locale

> * Translation file - Contains Message key and values for a given locale

Reviewer 2: ```suggestion
* Translation file - Contains Message keys and values for a given locale
```

> * Static Messages - Refers to the messages values stored in the translation file

Reviewer 2: This definition seems to be dependent upon the way the messages are stored rather than why they are static (and different to dynamic)

```suggestion
* Static (design / compile time) Messages - Message values that are defined during development and do not vary by deployment environment
```

* Dynamic Messages - Messages stored in the database that requires translation. A.k.a Runtime messages
* Message Format - Refers to the standard used for formatting the message values in the translation file

### In Scope Requirements/Use cases
- Return localized messages based on the value passed in the accept-language header
- Handle static messages with placeholder(s)

### Out of Scope Requirements/Use cases
- Naming convention to be used for keys
- Process for managing (decoupled from lokalise.com) translations
- Returning formatted (HTML/Markdown) messages
- Usage of [soft hyphen](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Softhyphentobreakwords) to break messages
- Support customization per tenant
- Process  for [back porting](https://wiki.folio.org/display/I18N/Backport) API/Backend messages similar to what we have for the front end 
- Support for [controlled vocabulary](https://issues.folio.org/browse/UXPROD-3148) (runtime data/data coming from DB tables. E.g. Patron Groups for tenants)

## Detailed Explanation/Design
#### API Protocol
* accept-language header value MUST be in ll_CC format. , where ll is a two-letter language code,

>   and CC is a two-letter country code.

Reviewer 2: I thought we agreed at the last meeting that we would adopt an external (IEEE?) standard for this rather than defining our own?

>   and CC is a two-letter country code. Refer to [IETF Language Tag standard](https://en.wikipedia.org/wiki/IETF_language_tag)

Reviewer: Why do we forbid tree letter IETF language codes like dsb, gsw, hsb, kok, moh, nso, prs, quc, quz, sah, sma, smj, smn, sms, syr, tzm?

>   and CC is a two-letter country code. Refer to [IETF Language Tag standard](https://en.wikipedia.org/wiki/IETF_language_tag)

Reviewer: Why is the country code required? In the standard it's optional, and FOLIO already as 18 translation files without country code, for example ja.json and de.json: https://github.com/folio-org/ui-users/tree/master/translations/ui-users

Reviewer 5: All of the questions about restrictions on the IETF Language Tags are good. If these restrictions are really meaningful to FOLIO, the RFC should say that FOLIO supports a subset of the IETF Language Tags. Or is it just that `ll-cc` is just the most familiar pattern?
Even though we may not have a use case today that requires then many possible levels of IETF tags, probably best to be open to the full IETF spec.

Reviewer 4: There are already different translations between `es-ES` (Spanish in Spain) versus `es-419`, so this distinction is meaningful today. (As I recall, Portuguese is another example where there are different translators working on different dialects...`pt-PT` (Portugual) versus `pt-BR` (Brazil).)

>   and CC is a two-letter country code. Refer to [IETF Language Tag standard](https://en.wikipedia.org/wiki/IETF_language_tag)

Reviewer 4: @vijay-eis : I think we can just refer to the IETF Language Tag standard and not try to describe what that standard is using this document.

>   and CC is a two-letter country code. Refer to [IETF Language Tag standard](https://en.wikipedia.org/wiki/IETF_language_tag)

Author: After having a quick discussion with @julianladisch  and @zburke here is what we have agreed to.

accept-language header value MUST be in languageCode-regionCode format. regionCode is an optional part of the value. **languageCode** is a two-letter code from [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1) (2002) or a three-letter code from [ISO 639-2](https://en.wikipedia.org/wiki/ISO_639-2) (1998), [ISO 639-3](https://en.wikipedia.org/wiki/ISO_639-3) (2007) or ISO 639-5 (2008). **regionCode** is  based on a two-letter country code from [ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) (written in upper case),

>   and CC is a two-letter country code. Refer to [IETF Language Tag standard](https://en.wikipedia.org/wiki/IETF_language_tag)

Reviewer 2: @vijay-eis 

> accept-language header value MUST be in languageCode-regionCode format. regionCode is an optional part of the value.

Is this consistent with or divergent from international standards?

It seems odd to me that we need to define our own concatenation of these. Is there on existing standard for how languages and regions are combined together?

>   and CC is a two-letter country code. Refer to [IETF Language Tag standard](https://en.wikipedia.org/wiki/IETF_language_tag)

Reviewer 2: > regionCode is based on a two-letter country code from [ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) (written in upper case),

Don't we have specific region codes that are not two letters?

The example at the start of this conversation uses {{419}} which would be non-compliant according to this description.

>   and CC is a two-letter country code. Refer to [IETF Language Tag standard](https://en.wikipedia.org/wiki/IETF_language_tag)

Reviewer: We took the first and forth bullet point from https://en.wikipedia.org/wiki/IETF_language_tag#Syntax_of_language_tags and removed "usually":

accept-language header value MUST be in languageCode-regionCode format, -regionCode is an optional part of the value.
* languageCode is a two-letter language code from [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1) (2002) or a three-letter code from [ISO 639-2](https://en.wikipedia.org/wiki/ISO_639-2) (1998), [ISO 639-3](https://en.wikipedia.org/wiki/ISO_639-3) (2007) or ISO 639-5 (2008), or registered through the BCP 47 process and composed of five to eight letters;
* regionCode is a two-letter country code from [ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) (written in upper case), or a three-digit code from [UN M.49](https://en.wikipedia.org/wiki/UN_M49) for geographical regions.

  MUST return the message in the en language.
* When handling messages with placeholders, replace the placeholders with the actual value on the server side before
    sending it back to the client
* Reliance on the lang query string parameter is DEPRECATED and SHOULD be updated to use the "Accept-language" header
  at the earliest possible development sprint.

#### Translation files
* Translation file format will depend on the language that is used to develop the module
  * Java, Groovy - properties
  * JavaScript, nodejs - json
  * Any other language - Please contact Technical Council for further guidance
* Translation files MUST be placed under translations/\<Backend Module Name\>, relative to the root of the repo.
  The files are being stored under a folder named after the module/repo to avoid having conflicts when using the same
  message key across multiple backend modules. 
* Translation file name MUST be [language code]_[country code].[extension] 
  where `language code` is an ISO 639 two-letter language code, `country code` is an ISO 3166 two-letter country code 
  and `extension` is the appropriate extension for the file format. For example, in `ll-CC.json`, 

>   ‘ll’ is the language code, ‘CC’ country code and `json` is the extension

Reviewer 2: I thought we agreed at the last meeting that we would adopt an external (IEEE?) standard for this rather than defining our own?

>   and `extension` is the appropriate extension for the file format.

Reviewer 3: We'll update the front-end to send language codes the back-end expects, i.e. language dash region instead of language underscore region.

>   and `extension` is the appropriate extension for the file format.

Reviewer: How is 419 converted into a two letter country code?
What country code is generated in cases like es.json or ja.json?

>   and `extension` is the appropriate extension for the file format.

Reviewer 4: @julianladisch : I'm not sure the source of your concern about two letter country codes. The part after the dash is a region (which in some cases is a country, but in other cases—like "419"—is bigger than a country). Do you have a use case where you need to guess a user's country by looking at the language code?

Aside, for @vijay-eis : I think we need to change [country code] to [region code] on line 75.

#### Translations
* Translation message keys MUST be a valid string 

> * Message Values in the translation file MUST be formatted according to the [ICU](https://icu.unicode.org)  standard

Reviewer 2: ```suggestion
* Message values in the translation file MUST be formatted according to the [ICU](https://icu.unicode.org)  standard
```

#### Language/Runtime
* Module SHOULD use language/framework specific recommended best practices for loading translation at runtime

> ## Rationale and Alternatives

Reviewer 2: I think the rationale (inc trade offs) for the chosen approach should be split from the other considered options (maybe they could be moved to an appendix).

It would make it more obvious to the reader which option was chosen (if it isn't apparent from the above) and that the trade offs had been considered. 

At the moment, the `Risks and Drawbacks` section is lost within these option and comes across to me like an after thought. 

I acknowledge that this is likely a side effect of this RFC being written after the decision was made. However I think if we also want this to be an exemplar of an RFC.

(I think these same challenges are apparent for the Kafka ADR as well, that is not currently a good example)
